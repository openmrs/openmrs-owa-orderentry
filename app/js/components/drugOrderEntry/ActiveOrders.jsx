import React from 'react';
import { connect } from 'react-redux';
import format from 'date-fns/format';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import activeOrderAction from '../../actions/activeOrderAction';
import { addDraftOrder } from '../../actions/draftTableAction';
import { setOrderAction } from '../../actions/orderAction';
import imageLoader from '../../../img/loading.gif';


export class ActiveOrders extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      draftOrder: {},
      pageCount: 0,
      limit: 10,
      startIndex: 0,
      pageNumber: 0,
      sort: 'desc',
    };
    this.onPageChange = this.onPageChange.bind(this);
  }
  componentDidMount() {
    this.fetchActiveOrders(this.props);
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ pageCount: nextProps.pageCount });
  }

  onClickDiscontinue = (order) => {
    const {
      uuid,
      drug,
      dose,
      doseUnits,
      frequency,
      route,
      duration,
      careSetting,
      durationUnits,
      asNeededCondition,
      orderer,
      dosingInstructions,
      dosingType,
      type,
      quantity,
      quantityUnits,
      orderNumber,
    } = order;

    this.setState({
      draftOrder: {
        uuid,
        careSetting: careSetting.uuid,
        drugName: drug.display,
        action: 'DISCONTINUE',
        dose,
        drug: drug.uuid,
        dosingUnit: (doseUnits && doseUnits.display) || doseUnits,
        frequency: (frequency && frequency.display) || frequency,
        route: (route && route.display) || route,
        duration,
        type,
        orderer: orderer.uuid,
        durationUnit: (durationUnits && durationUnits.display) || durationUnits,
        dosingType,
        reason: asNeededCondition,
        drugInstructions: dosingInstructions,
        dispensingQuantity: quantity,
        dispensingUnit: (quantityUnits && quantityUnits.display) || quantityUnits,
        orderNumber,
      },
    }, () => {
      this.props.setOrderAction('DISCONTINUE', orderNumber);
      this.props.addDraftOrder(this.state.draftOrder);
    });
  }

  onPageChange({ selected }) {
    let { startIndex, pageNumber } = this.state;
    const { limit } = this.state;
    startIndex = Math.ceil(selected * limit);
    pageNumber = startIndex / limit;
    this.setState(() => ({ startIndex, pageNumber }));
    this.fetchActiveOrders(this.props, { limit, startIndex });
  }

  fetchActiveOrders(props, newState) {
    const { limit, startIndex = 0 } = newState || this.state;
    this.setState(() => ({ startIndex: 0 }));
    const { location, careSetting } = props;
    const query = new URLSearchParams(location.search);
    const patientUuid = query.get('patient');
    const caresettingUuid = careSetting.uuid;
    this.props.activeOrderAction(limit, startIndex, patientUuid, caresettingUuid, this.state.sort);
  }

  showOrders = activeOrders => activeOrders
    .map((order) => {
      const {
        uuid,
        action,
        careSetting,
        dateActivated,
        autoExpireDate,
        drug,
        dose,
        doseUnits,
        frequency,
        route,
        duration,
        durationUnits,
        dosingInstructions,
        quantity,
        quantityUnits,
        dosingType,
        orderer,
        orderNumber,
        status,
      } = order;

      let editClass;
      if (status === 'EDIT' || status === 'DRAFT_EDIT') editClass = 'current-order-color';

      let details;
      if (dosingType === 'org.openmrs.SimpleDosingInstructions') {
        details = (
          <td>

            {drug.display}:
            {` ${dose}`}
            {` ${doseUnits.display}`}
            {`, ${frequency.display}`}
            {`, ${route.display}`}
            {duration && `, for ${duration}`}
            {durationUnits && ` ${durationUnits.display} total`}
            {dosingInstructions && `, (${dosingInstructions})`}
            {(quantity && quantityUnits) && `, (Dispense: ${quantity} ${quantityUnits.display})`}

          </td>
        );
      } else {
        details = (
          <td className={editClass}>

            {drug.display}:
            {dosingInstructions && ` ${dosingInstructions}`}
            {(quantity && quantityUnits) && `, (Dispense: ${quantity} ${quantityUnits.display})`}

          </td>
        );
      }

      let showStatus;

      if (status === 'EDIT' || status === 'DRAFT_EDIT') {
        showStatus = (
          <p> will REVISE </p>
        );
      } else if (status === 'DISCONTINUE') {
        showStatus = (
          <p> will DISCONTINUE </p>
        );
      } else {
        showStatus = (
          <div>
            <a
              id="edit-drug-orders"
              href="#"
              onClick={() => this.props.handleEditActiveDrugOrder(order, details)}
            > <i className="icon-edit" title="Edit" />
            </a>
            <a > <i
              className="icon-remove icon-color"
              title="Discontinue"
              id="delete"
              role="button"
              tabIndex="0"
              onKeyPress={this.handleKeyPress}
              onClick={() => this.onClickDiscontinue(order)} />
            </a>
          </div>
        );
      }

      const { dateFormat } = this.props;

      return (
        <tr key={uuid} >
          <td id="table-font" className={editClass}>
            {format(dateActivated, dateFormat)}
            {autoExpireDate && (`- ${format(autoExpireDate, dateFormat)}`)}
          </td>
          {details}
          <td className="text-center action-padding" >
            {showStatus}
          </td>
        </tr>
      );
    })


  render() {
    const { activeOrders, loading } = this.props.drugOrder;
    const { activity, selectedOrder } = this.props;
    let allActiveOrders = activeOrders;
    if (activity === 'EDIT') {
      allActiveOrders = [
        selectedOrder,
        ...activeOrders.filter(order => order.orderNumber !== selectedOrder.orderNumber),
      ];
    }

    if (!activeOrders || loading) {
      return (
        <div className="text-align-center">
          <img src={imageLoader} alt="loader" />
        </div>
      );
    } else if (activeOrders.length === 0) {
      return (
        <div>
          <p>No Active Orders</p>
        </div>
      );
    }
    return (
      <div>
        <table className="table bordered mw-958-px">
          <thead>
            <tr>
              <th className="w-155-px">Date</th>
              <th>Details</th>
              <th className="w-81-px">Actions</th>
            </tr>
          </thead>
          <tbody>
            {this.showOrders(allActiveOrders)}
          </tbody>
        </table>
        {
          this.props.pageCount !== 1 &&
            <div className="clear">
              <div className="float-left-padding">
                {this.props.showResultCount}
              </div>
              <div className="dataTables_paginate">
                <ReactPaginate
                  pageCount={this.state.pageCount}
                  pageRangeDisplayed={5}
                  marginPagesDisplayed={3}
                  previousLabel="Previous"
                  nextLabel="Next"
                  breakClassName="text-align-center"
                  initialPage={0}
                  containerClassName="react-paginate-container"
                  pageLinkClassName="page-link"
                  activeClassName="active-link"
                  disabledClassName="active-link"
                  nextLinkClassName="page-link"
                  previousLinkClassName="page-link"
                  onPageChange={this.onPageChange}
                  forcePage={this.state.pageNumber}
                  disableInitialCallback
                />
              </div>
            </div>
        }
      </div>
    );
  }
}

ActiveOrders.defaultProps = {
  showResultCount: 'Showing 1 to 10 of 55 entries',
  pageCount: 0,

};


const mapStateToProps = ({ activeOrderReducer, orderSelectionReducer }) => ({
  drugOrder: activeOrderReducer,
  pageCount: activeOrderReducer.pageCount,
  showResultCount: activeOrderReducer.showResultCount,
  selectedOrder: orderSelectionReducer.selectedOrder,
  activity: orderSelectionReducer.activity,
});

const actionCreators = {
  activeOrderAction,
  addDraftOrder,
  setOrderAction,
};

ActiveOrders.propTypes = {
  handleEditActiveDrugOrder: PropTypes.func.isRequired,
  activeOrderAction: PropTypes.func.isRequired,
  dateFormat: PropTypes.string.isRequired,
  addDraftOrder: PropTypes.func.isRequired,
  setOrderAction: PropTypes.func.isRequired,
  drugOrder: PropTypes.shape({
    loading: PropTypes.bool,
    activeOrders: PropTypes.arrayOf(PropTypes.shape({})),
  }),
  activeOrders: PropTypes.shape({}).isRequired,
  showResultCount: PropTypes.string,
  pageCount: PropTypes.number,
  selectedOrder: PropTypes.object,
  activity: PropTypes.string,
};

ActiveOrders.defaultProps = {
  drugOrder: {
    loading: false,
    activeOrders: [],
  },
  selectedOrder: {},
  activity: '',
};

export default connect(mapStateToProps, actionCreators)(ActiveOrders);
