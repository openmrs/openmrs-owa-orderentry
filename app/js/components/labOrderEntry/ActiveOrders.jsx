import React from 'react';
import PropTypes from 'prop-types';

const ActiveOrders = ({ labOrderData, tests }) => (
  <div>
    <table className="table bordered mw-958-px">
      <thead>
        <tr>
          <th>Name</th>
          <th>tests</th>
          <th>status</th>
        </tr>
      </thead>
      <tbody>
        <tr key={labOrderData.uuid}>
          <td>{labOrderData.display}</td>
          <td>{tests.join(', ')}</td>
          <td>awaiting-results</td>
        </tr>
      </tbody>
    </table>
  </div>
);

export default ActiveOrders;

ActiveOrders.defaultProps = {
  labOrderData: {
    uuid: '',
    display: '',
  },
  tests: [''],
};

ActiveOrders.propTypes = {
  labOrderData: PropTypes.shape({
    uuid: PropTypes.string,
    display: PropTypes.string,
  }),
  tests: PropTypes.arrayOf(PropTypes.string),
};
