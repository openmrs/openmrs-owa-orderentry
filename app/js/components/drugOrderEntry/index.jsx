import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Accordion } from '@openmrs/react-components';
import AddForm from './addForm/AddForm';
import PastOrders from './PastOrders';
import Tabs from '../tabs/Tabs';
import Tab from '../tabs/Tab';
import SearchDrug from '../searchDrug';
import ActiveOrders from './ActiveOrders';
import { setOrderAction } from '../../actions/orderAction';
import { deleteDraftOrder, deleteAllDraftOrders } from '../../actions/draftTableAction';
import DraftDataTable from './addForm/DraftDataTable';
import { selectDrugSuccess } from '../../actions/drug';
import './styles.scss';

export class SearchAndAddOrder extends React.PureComponent {
  state = {
    value: "",
    focused: false,
    editDrugUuid: '',
    editDrugName: '',
    editOrder: {},
    draftOrder: {},
  };

  componentDidMount() {
    const { selectedOrder, activity } = this.props;
    if (activity === 'EDIT' && !this.state.formattedDetails) {
      const details = (
        <td>
          {selectedOrder.drug.display}:
          {selectedOrder.dosingInstructions && ` ${selectedOrder.dosingInstructions}`}
          {(selectedOrder.quantity && selectedOrder.quantityUnits) && `, (Dispense: ${selectedOrder.quantity} ${selectedOrder.quantityUnits.display})`}
        </td>
      );
      this.handleEditActiveDrugOrder({ ...selectedOrder, status: activity }, details);
    }
  }

  onSelectDrug = (drugName) => {
    this.setState(() => ({
      value: drugName,
      focused: false,
    }));
  }

  onChange = (value) => {
    this.setState(() => ({
      value,
      focused: true,
    }));
  }

  handleDiscardOneOrder = (order) => {
    this.props.deleteDraftOrder(order);
    if (order.action === 'REVISE') {
      this.props.setOrderAction('DISCARD_EDIT', order.orderNumber);
    } else {
      this.props.setOrderAction('DISCARD_ONE', order.orderNumber);
    }
  }

  handleDiscardAllOrders = () => {
    this.props.deleteAllDraftOrders();
    this.props.setOrderAction('DISCARD_ALL', '0');
  }

  clearSearchField = () => {
    this.setState({
      value: "",
      focused: false,
      editDrugUuid: "",
      editDrugName: "",
    });
  }

  handleEditDraftOrder = (order) => {
    this.props.selectDrugSuccess(order.drug);
    this.setState({
      draftOrder: order,
      editDrugUuid: order.drug,
      editDrugName: order.drugName,
      orderNumber: order.orderNumber,
    }, () => {
      this.props.setOrderAction('EDIT', order.orderNumber);
    });
    this.props.deleteDraftOrder(order);
  }

  handleEditActiveDrugOrder = (order, details) => {
    let formattedDetails = details.props.children.join("");
    if (formattedDetails.length > 250) {
      formattedDetails = `${formattedDetails.substring(0, 250)}...`;
    }
    const editOrder = order;
    editOrder.action = "REVISE";
    this.setState({
      formattedDetails,
      editDrugUuid: order.drug.uuid,
      editDrugName: order.drug.display,
      editOrder,
      orderNumber: order.orderNumber,
    }, () => {
      this.props.setOrderAction('EDIT', order.orderNumber);
    });
  }

  removeOrder = () => {
    this.setState({
      editOrder: {},
      draftOrder: {},
    });
  }

  closeFormsOnTabChange = () => {
    this.clearSearchField();
    this.props.selectDrugSuccess();
    this.props.deleteAllDraftOrders();
  }

  renderSearchDrug = () => (
    this.state.editDrugName ?
      <p className="revise-order-padding">
        <b className="revised-order-text"> Revise for: {this.state.editDrugName} </b><br />
        <b className="current-order-color">Current Order: <em>{this.state.formattedDetails}</em></b>
      </p>
      :
      <SearchDrug
        value={this.state.value}
        focused={this.state.focused}
        onChange={this.onChange}
        onSelectDrug={this.onSelectDrug}
      />
  )

  renderAddForm = careSetting => (
    <div>
      <AddForm
        drugName={this.state.editDrugName ? this.state.editDrugName : this.props.drug.display}
        drugUuid={this.state.editDrugUuid ? this.state.editDrugUuid : this.props.drug.uuid}
        editOrder={this.state.editOrder}
        careSetting={careSetting}
        clearSearchField={this.clearSearchField}
        removeOrder={this.removeOrder}
        draftOrder={this.state.draftOrder}
        orderNumber={this.state.orderNumber}
      />
    </div>
  );

  renderDraftDataTable = careSetting => (
    (this.props.draftOrders.length > 0) &&
    <DraftDataTable
      draftOrders={this.props.draftOrders}
      handleDiscardOneOrder={this.handleDiscardOneOrder}
      handleDiscardAllOrders={this.handleDiscardAllOrders}
      handleEditDraftOrder={this.handleEditDraftOrder}
      careSetting={careSetting}
    />
  )

  render() {
    const {
      outpatientCareSetting, location, dateFormat,
    } = this.props;
    return (
      <div className="drug-order-entry">
        <h5 className="drug-form-header">New Drug Order</h5>
        {this.renderSearchDrug()}
        {this.renderAddForm(outpatientCareSetting)}
        {this.renderDraftDataTable(outpatientCareSetting)}
        <Accordion open title="Active Drug Orders">
          <ActiveOrders
            careSetting={outpatientCareSetting}
            location={location}
            dateFormat={dateFormat}
            handleEditActiveDrugOrder={this.handleEditActiveDrugOrder}
          />
        </Accordion>

        <Accordion title="Past Drug Orders">
          <PastOrders
            careSetting={outpatientCareSetting}
            dateFormat={dateFormat}
            location={location}
          />
          <br />
        </Accordion>
      </div>
    );
  }
}

const mapStateToProps = ({
  careSettingReducer:
  { outpatientCareSetting },
  drugSearchReducer,
  draftReducer: { draftDrugOrders },
  dateFormatReducer: { dateFormat },
  orderSelectionReducer: { activity, selectedOrder },
}) => ({
  outpatientCareSetting,
  drug: drugSearchReducer.selected,
  draftOrders: draftDrugOrders,
  dateFormat,
  activity,
  selectedOrder,
});

SearchAndAddOrder.defaultProps = {
  draftOrders: [],
  drug: null,
  selectedOrder: {},
  activity: '',
};

SearchAndAddOrder.propTypes = {
  drug: PropTypes.oneOfType([
    PropTypes.shape({
      uuid: PropTypes.string,
      display: PropTypes.string,
    }),
    PropTypes.string,
  ]),
  draftOrders: PropTypes.arrayOf(PropTypes.any),
  dateFormat: PropTypes.string.isRequired,
  location: PropTypes.shape({
    search: PropTypes.string,
  }).isRequired,
  setOrderAction: PropTypes.func.isRequired,
  deleteDraftOrder: PropTypes.func.isRequired,
  deleteAllDraftOrders: PropTypes.func.isRequired,
  outpatientCareSetting: PropTypes.shape({
    uuid: PropTypes.string.isRequired,
    display: PropTypes.string.isRequired,
  }).isRequired,
  selectDrugSuccess: PropTypes.func.isRequired,
  selectedOrder: PropTypes.object,
  activity: PropTypes.string,
};

export default connect(
  mapStateToProps,
  {
    setOrderAction,
    deleteDraftOrder,
    deleteAllDraftOrders,
    selectDrugSuccess,
  },
)(SearchAndAddOrder);
