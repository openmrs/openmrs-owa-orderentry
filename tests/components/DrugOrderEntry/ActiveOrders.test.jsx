import React from 'react';

import { ActiveOrders } from '../../../app/js/components/drugOrderEntry/ActiveOrders';
import { AddForm } from '../../../app/js/components/drugOrderEntry/addForm/AddForm';

const { data, careSetting, order } = mockData;

const props = {
  handleEditActiveDrugOrder: jest.fn(),
  activeOrderAction: jest.fn(),
  addDraftOrder: jest.fn(),
  onDelete: jest.fn(),
  setOrderAction: jest.fn(),
  isDelete: false,
  drugOrder: {
    activeOrders: [
      {
      drug: { display: 'Morphine'},
      uuid:'024b1459-f31d-46cb-8ad3-2a624c894e2c',
      dose:1,
      quantity:2,
      duration:2,
      auditInfo:{dateCreated:'2018-03-20 10:59:22'},
      drug:{display:'new drug'},
      doseUnits:{display:'drops'},
      frequency:{display:'twice a week'},
      route:{display:'reactally'},
      quantityUnits:{display:'drops'},
      durationUnits:{display:'weeks'},
      dosingType: "org.openmrs.SimpleDosingInstructions",
      dosingInstructions: '3 tabs'
      }
    ],
    loading: false
  },
  dateFormat: 'DD-MMM-YYYY HH:mm',
  location: {
    search: {}
  },
  careSetting,
  data,
  order,
  tabName: '',
  activeOrders: {}
}

let mountedComponent;
const getComponent = () => {
  if (!mountedComponent) {
    mountedComponent = shallow(<ActiveOrders {...props} store={store} />);
  }
  return mountedComponent;
};

describe('Test for Inpatient Active orders', () => {
  it('should render component', () => {
    expect(getComponent()).toMatchSnapshot()
  });
});

describe('onClickDiscontinue() method', () => {
  it('should call onClickDiscontinue()', () => {
    const renderedComponent = getComponent().instance();
    sinon.spy(renderedComponent, 'onClickDiscontinue');
    renderedComponent.onClickDiscontinue(order);
    expect(renderedComponent.onClickDiscontinue.calledOnce).toEqual(true);
    expect(getComponent().state('draftOrder')).toEqual({
      action: "DISCONTINUE",
      careSetting: "",
      dispensingQuantity: undefined,
      dispensingUnit: undefined,
      dose: "",
      dosingUnit: "",
      dosingType: "org.openmrs.SimpleDosingInstructions",
      dosingUnit: undefined,
      drug: "",
      drugInstructions: undefined,
      duration: "",
      durationUnit: undefined,
      frequency: "",
      reason: undefined,
      route: "",
      orderNumber: "",
      orderer: "",
      type: undefined,
      uuid: "",
      drugName: "",
    });
  });
});

describe('fetchActiveOrders() method', () => {
  it('should call fetchActiveOrders()', () => {
    const renderedComponent = getComponent().instance();
    sinon.spy(renderedComponent, 'fetchActiveOrders');
    renderedComponent.fetchActiveOrders(props);
    expect(renderedComponent.fetchActiveOrders.calledOnce).toEqual(true);
  });
});

describe('onPageChange() method', () => {
  it('should call onPageChange()', () => {
    const renderedComponent = getComponent().instance();
    sinon.spy(renderedComponent, 'onPageChange');
    renderedComponent.onPageChange(data);
    expect(renderedComponent.onPageChange.calledOnce).toEqual(true);
  });
});

describe('behaviour when dosing type is free text', () => {
  it('should still display a table', () => {
    const props = {
      handleEditActiveDrugOrder: jest.fn(),
      activeOrderAction: jest.fn(),
      addDraftOrder: jest.fn(),
      onDelete: jest.fn(),
      setOrderAction: jest.fn(),
      isDelete: false,
      drugOrder: {
        activeOrders: [
          {
            drug: { display: 'Morphine'},
            uuid:'024b1459-f31d-46cb-8ad3-2a624c894e2c',
            dose:1,
            quantity:2,
            duration:2,
            auditInfo:{dateCreated:'2018-03-20 10:59:22'},
            drug:{display:'new drug'},
            doseUnits:{display:'drops'},
            frequency:{display:'twice a week'},
            route:{display:'reactally'},
            quantityUnits:{display:'drops'},
            durationUnits:{display:'weeks'},
            dosingType: "org.openmrs.FreeTextInstructions",
            }
        ],
        loading: false
      },
      location: {
        search: {}
      },
      careSetting,
      data,
      order: {},
      tabName: '',
      activeOrders: {},
      dateFormat: ''
    }
    const wrapper = shallow(<ActiveOrders {...props} />);
    expect(wrapper.find('table').length).toEqual(1);
  });
});

describe('behaviour when status is edit', () => {
  it('should display  <p> will REVISE </p>', () => {
    const props = {
      handleEditActiveDrugOrder: jest.fn(),
      activeOrderAction: jest.fn(),
      addDraftOrder: jest.fn(),
      onDelete: jest.fn(),
      setOrderAction: jest.fn(),
      isDelete: false,
      drugOrder: {
        activeOrders: [
          {
            drug: { display: 'Morphine'},
            uuid:'024b1459-f31d-46cb-8ad3-2a624c894e2c',
            dose:1,
            quantity:2,
            duration:2,
            auditInfo:{dateCreated:'2018-03-20 10:59:22'},
            drug:{display:'new drug'},
            doseUnits:{display:'drops'},
            frequency:{display:'twice a week'},
            route:{display:'reactally'},
            quantityUnits:{display:'drops'},
            durationUnits:{display:'weeks'},
            dosingType: "org.openmrs.FreeTextInstructions",
            status:'EDIT'
            }
        ],
        loading: false
      },
      location: {
        search: {}
      },
      careSetting,
      data,
      order: {
        status: 'EDIT'
      },
      tabName: '',
      activeOrders: {},
      dateFormat: '',
    }
    const mockCallBack = jest.fn();
    const wrapper = shallow(<ActiveOrders onClick={mockCallBack} {...props} />);
    const renderedComponent = getComponent().instance();
    sinon.spy(renderedComponent, 'showOrders');
    renderedComponent.showOrders(props.drugOrder.activeOrders);
    expect(renderedComponent.showOrders.calledOnce).toEqual(true);
    expect(wrapper.find('table').length).toEqual(1);
  });

  it('should call edit', () => {
    const props = {
      handleEditActiveDrugOrder: jest.fn(),
      activeOrderAction: jest.fn(),
      addDraftOrder: jest.fn(),
      onDelete: jest.fn(),
      setOrderAction: jest.fn(),
      isDelete: false,
      drugOrder: {
        activeOrders: [
          {
            drug: { display: 'Morphine'},
            uuid:'024b1459-f31d-46cb-8ad3-2a624c894e2c',
            dose:1,
            quantity:2,
            duration:2,
            auditInfo:{dateCreated:'2018-03-20 10:59:22'},
            drug:{display:'new drug'},
            doseUnits:{display:'drops'},
            frequency:{display:'twice a week'},
            route:{display:'reactally'},
            quantityUnits:{display:'drops'},
            durationUnits:{display:'weeks'},
            dosingType: "org.openmrs.FreeTextInstructions",
            }
        ],
        loading: false
      },
      location: {
        search: {}
      },
      careSetting,
      data,
      tabName: '',
      activeOrders: {},
      dateFormat: '',
    }
    const mockCallBack = jest.fn();
    const wrapper = shallow(<ActiveOrders onClick={mockCallBack} {...props} />);
    wrapper.find('a').find('#edit-drug-orders').simulate('click');
  });
});

describe('behaviour when status is DISCONTINUE', () => {
  it('should display  <p> will DISCONTINUE </p>', () => {
    const props = {
      handleEditActiveDrugOrder: jest.fn(),
      activeOrderAction: jest.fn(),
      addDraftOrder: jest.fn(),
      onDelete: jest.fn(),
      setOrderAction: jest.fn(),
      isDelete: true,
      drugOrder: {
        activeOrders: [
          {
            drug: { display: 'Morphine'},
            uuid:'024b1459-f31d-46cb-8ad3-2a624c894e2c',
            dose:1,
            quantity:2,
            duration:2,
            auditInfo:{dateCreated:'2018-03-20 10:59:22'},
            drug:{display:'new drug'},
            doseUnits:{display:'drops'},
            frequency:{display:'twice a week'},
            route:{display:'reactally'},
            quantityUnits:{display:'drops'},
            durationUnits:{display:'weeks'},
            dosingType: "org.openmrs.FreeTextInstructions",
            careSetting: {uuid: ''},
            orderer: {uuid: ''},
            status:"DISCONTINUE"
            }
        ],
        loading: false
      },
      location: {
        search: {}
      },
      careSetting,
      data,
      order: {
        status: 'DISCONTINUE'
      },
      tabName: '',
      activeOrders: {},
      dateFormat: '',
    }
    const mockCallBack = jest.fn();
    const wrapper = shallow(<ActiveOrders onClick={mockCallBack} {...props} />);
    expect(wrapper.find('table').length).toEqual(1);
  });

  it('should call delete', () => {
    const props = {
      handleEditActiveDrugOrder: jest.fn(),
      activeOrderAction: jest.fn(),
      addDraftOrder: jest.fn(),
      onDelete: jest.fn(),
      setOrderAction: jest.fn(),
      isDelete: true,
      drugOrder: {
        activeOrders: [
          {
            drug: { display: 'Morphine'},
            uuid:'024b1459-f31d-46cb-8ad3-2a624c894e2c',
            dose:1,
            quantity:2,
            duration:2,
            auditInfo:{dateCreated:'2018-03-20 10:59:22'},
            drug:{display:'new drug'},
            doseUnits:{display:'drops'},
            frequency:{display:'twice a week'},
            route:{display:'reactally'},
            quantityUnits:{display:'drops'},
            durationUnits:{display:'weeks'},
            dosingType: "org.openmrs.FreeTextInstructions",
            careSetting: {uuid: ''},
            orderer: {uuid: ''},
            }
        ],
        loading: false
      },
      location: {
        search: {}
      },
      careSetting,
      data,
      tabName: '',
      activeOrders: {},
      dateFormat: ''
    }
    const mockCallBack = jest.fn();
    const wrapper = shallow(<ActiveOrders onClick={mockCallBack} {...props} />);
    wrapper.find('a').find('#delete').simulate('click');
  })
});


describe('behaviour when there is no active orders', () => {
  it('should display <p>No Active Orders</p>', () => {
    const props = {
      handleEditActiveDrugOrder: jest.fn(),
      activeOrderAction: jest.fn(),
      addDraftOrder: jest.fn(),
      onDelete: jest.fn(),
      setOrderAction: jest.fn(),
      isDelete: false,
      drugOrder: {
        activeOrders: [],
        loading: false
      },
      location: {
        search: {}
      },
      careSetting,
      data,
      order,
      tabName: '',
      activeOrders: {},
      dateFormat: ''
    }
    const wrapper = shallow(<ActiveOrders {...props} />);
    expect(wrapper.find('p').text()).toEqual('No Active Orders');
  });
});

describe('behaviour when there is no active orders and page is loading', () => {
  it('should display <img src={imageLoader} alt="loader" />', () => {
    const props = {
      handleEditActiveDrugOrder: jest.fn(),
      activeOrderAction: jest.fn(),
      addDraftOrder: jest.fn(),
      onDelete: jest.fn(),
      setOrderAction: jest.fn(),
      isDelete: false,
      drugOrder: {
        activeOrders: [],
        loading: true
      },
      location: {
        search: {}
      },
      careSetting,
      data,
      order,
      tabName: '',
      activeOrders: {},
      dateFormat: ''
    }
    const wrapper = shallow(<ActiveOrders {...props} />);
    expect(wrapper.find('img').length).toEqual(1);
  });
});

describe('componentWillReceiveProps()', () => {
  it('call fetchActiveOrders and udpate state', () => {
      const renderedComponent = getComponent().instance();
      getComponent().setProps({ tabName: 'outpatient' })
      expect(renderedComponent.fetchActiveOrders.calledOnce).toEqual(false);
      getComponent().setProps({ pageCount: 9 })
      expect(getComponent().state().pageCount).toEqual(9);
  });
});
