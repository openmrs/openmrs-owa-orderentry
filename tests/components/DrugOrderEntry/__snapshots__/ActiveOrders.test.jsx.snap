// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for Inpatient Active orders should render component 1`] = `
<div>
  <table
    className="table bordered mw-958-px"
  >
    <thead>
      <tr>
        <th
          className="w-155-px"
        >
          Date
        </th>
        <th>
          Details
        </th>
        <th
          className="w-81-px"
        >
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="024b1459-f31d-46cb-8ad3-2a624c894e2c"
      >
        <td
          id="table-font"
        >
          Invalid Date
        </td>
        <td>
          new drug
          :
           1
           drops
          , twice a week
          , reactally
          , for 2
           weeks total
          , (3 tabs)
          , (Dispense: 2 drops)
        </td>
        <td
          className="text-center action-padding"
        >
          <div>
            <a
              href="#"
              id="edit-drug-orders"
              onClick={[Function]}
            >
               
              <i
                className="icon-edit"
                title="Edit"
              />
            </a>
            <a>
               
              <i
                className="icon-remove icon-color"
                id="delete"
                onClick={[Function]}
                role="button"
                tabIndex="0"
                title="Discontinue"
              />
            </a>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="clear"
  >
    <div
      className="float-left-padding"
    />
    <div
      className="dataTables_paginate"
    >
      <PaginationBoxView
        activeClassName="active-link"
        breakClassName="text-align-center"
        breakLabel="..."
        containerClassName="react-paginate-container"
        disableInitialCallback={true}
        disabledClassName="active-link"
        forcePage={0}
        initialPage={0}
        marginPagesDisplayed={3}
        nextClassName="next"
        nextLabel="Next"
        nextLinkClassName="page-link"
        onPageChange={[Function]}
        pageCount={0}
        pageLinkClassName="page-link"
        pageRangeDisplayed={5}
        previousClassName="previous"
        previousLabel="Previous"
        previousLinkClassName="page-link"
      />
    </div>
  </div>
</div>
`;
